src/cheri_ptw.sail

```diff
@@ -24,22 +24,58 @@ function ptw_error_to_str(e) =
 
 overload to_str = {ptw_error_to_str}
 
+/*
+ * I apologize for the confusion you have experienced if you have skipped over
+ * this comment and read the body of this function first.  Hopefully these two
+ * points resolve most of it:
+ *
+ * 1) While it may look like we're signaling that we must *always* raise
+ * faults, translationException below can see whether the instruction
+ * triggering the PTW is claiming to store an asserted tag, and will ignore our
+ * result if that isn't so.
+ *
+ * 2) We analyse only the store-side component of ext_ptw here.  The PTW logic
+ * cannot see into the data path, and so load-side responses, including traps,
+ * must be handled by the instruction that initiated the translation.
+ */
 function ext_get_ptw_error(ext_ptw : ext_ptw) -> PTW_Error =
-  match (ext_ptw) {
-    PTW_CAP_OK        => PTW_No_Permission(),
-    PTW_LOAD_CAP_ERR  => PTW_Ext_Error(AT_CAP_ERR),
-    PTW_STORE_CAP_ERR => PTW_Ext_Error(AT_CAP_ERR)
+  match (ext_ptw.ptw_sc_mod) {
+
+    /*
+     * This is confusing: PTW_No_Permission() is never used as such, but just
+     * dodges the pattern matches in translationException, below.
+     */
+    PTW_SCM_OK   => PTW_No_Permission(),
+
+    PTW_SCM_TRAP => PTW_Ext_Error(AT_CAP_ERR)
   }
 
 /* conversion of these translation/PTW failures into architectural exceptions */
 function translationException(a : AccessType(ext_access_type), f : PTW_Error) -> ExceptionType = {
   let e : ExceptionType =
   match (a, f) {
-    /* First handle the vmem extension errors */
-    (ReadWrite(Cap), PTW_Ext_Error(AT_CAP_ERR)) => E_Extension(EXC_SAMO_CAP_PAGE_FAULT),
-    (Read(Cap), PTW_Ext_Error(AT_CAP_ERR))      => E_Extension(EXC_LOAD_CAP_PAGE_FAULT),
+
+    /* Tag-asserting stores can raise CHERI page faults. */
     (Write(Cap), PTW_Ext_Error(AT_CAP_ERR))     => E_Extension(EXC_SAMO_CAP_PAGE_FAULT),
 
+    /*
+     * XXX ReadWrite(Cap) does not provide enough information to distinguish
+     * between a tag-clearing store and tag-asserting store in the presence of
+     * a tag-capable load, or a tag-asserting store in the case of a
+     * tag-incapable load.  (That is, it only clearly distinguishes a
+     * tag-incapable load with a tag-clearing store from every other option.)
+     * Groundwork for fixing this exists in the sail RISC-V model, but we have
+     * not yet caught up to that point.  Fortunately, we don't actually have
+     * amoswap.c yet, so this is not observably wrong.
+     */
+    (ReadWrite(Cap), PTW_Ext_Error(AT_CAP_ERR)) => E_Extension(EXC_SAMO_CAP_PAGE_FAULT),
```

Comment:
Would you be willing to update us to the latest sail-riscv and fix this in a subsequent PR? :)